### PR TITLE
Restructure web service driver and event firing driver.

### LINKED ISSUE #390: TaskCanceledException when using PostWithResponse
hi,

this was supposed to be an update on #387 but as it's now closed i decided to report as a new ticket.

after modifying the test to follow the best practices i noticed the following error and it's pretty consistent. from what i googled, when consuming a `HttpResponseMessage` i should get the response using `ReadAsStringAsync()` or `ReadAsStreamAsync()` otherwise i can get this `TaskCanceledException`.

the exception
```
System.Threading.Tasks.TaskCanceledException : The operation was canceled.
  ----> System.IO.IOException : Unable to read data from the transport connection: Operation canceled.
  ----> System.Net.Sockets.SocketException : Operation canceled
```

here is the reworked test
```
        const string contentTypeJson = "application/json";

        [Test]
        public void CreateNewPost()
        {
            // Arrange
            var newPost = new Post("my title", "my description", 5); // this is a simple POJO, nothing special
            var content = WebServiceUtils.MakeStringContent<Post>(newPost, Encoding.UTF8, contentTypeJson);

            // Action
            var response = this.WebServiceDriver.PostWithResponse("/posts", contentTypeJson, content);

            // Assert
            Assert.AreEqual(response.StatusCode, HttpStatusCode.Created);

            // if i use this it always work
            //Task<string> body = response.Content.ReadAsStringAsync();
            //var sBody = body.Result;

            Post responsePost = WebServiceUtils.DeserializeJson<Post>(response);
            // shortcut so i don't need to implement object comparison
            Assert.AreEqual(newPost.title, responsePost.title);
            Assert.AreEqual(newPost.body, responsePost.body);
            Assert.AreEqual(newPost.userId, responsePost.userId);
            Assert.AreEqual(newPost.id, 101);
        }
```

### LINKED ISSUE #646: PatchWithResponse does not log with EventFiringWebServiceDriver
With the config variable Log set to OnFail or Yes, PatchWithResponse does not log to the logger.

```
var content = WebServiceUtils.MakeStringContent(patchData, Encoding.UTF8, Constants.MediaTypeJson);
var updateResponse = this.WebServiceDriver.PatchWithResponse(updateUrl, Constants.MediaTypeJson, content, false);
```

### PR DESCRIPTION
**Pull request by TroyWalshProf**
Make sure the event firing driver logs for each type of service call.
Check of and note service call timeouts.

Fixes #646
Fixes #390

### CODE REVIEW

File: Framework/BaseWebServiceTest/EventFiringWebServiceDriver.cs [hunk 1]
```diff
@@ -294,5 +255,38 @@ private void RaiseErrorMessage(Exception e)
...
+                    }
+                }
+                else
+                {
+                    message.AppendLine("  **Writting this kind of content to the log is not supported**");
+                }
+            }
+            else
+            {
+                message.AppendLine("  **Content is NULL**");
```
Comment (FermJacob) [inline Framework/BaseWebServiceTest/EventFiringWebServiceDriver.cs:288]: This should be mediatype is null or empty

File: Framework/BaseWebServiceTest/EventFiringWebServiceDriver.cs [hunk 2]
```diff
@@ -294,5 +255,38 @@ private void RaiseErrorMessage(Exception e)
...
+        /// </summary>
+        /// <param name="message">String builder for building the message</param>
+        /// <param name="mediaType">Content media type</param>
+        /// <param name="content">The content we are building a message for</param>
+        private void BuildContentMessage(StringBuilder message, string mediaType, HttpContent content)
+        {
+            message.AppendLine("Content:");
+            message.AppendLine($"  Content Media Type: {mediaType}");
+            message.AppendLine("  Content Text:");
+            if (mediaType != null)
```
Comment (FermJacob) [inline Framework/BaseWebServiceTest/EventFiringWebServiceDriver.cs:270]: Should we perform string.IsNullOrEmpty check instead?

File: Framework/BaseWebServiceTest/EventFiringWebServiceDriver.cs [hunk 3]
```diff
@@ -294,5 +255,38 @@ private void RaiseErrorMessage(Exception e)
...
+            { 
+                mediaType = mediaType.ToUpper();
+                
+                if (mediaType.Contains("TEXT") || mediaType.Contains("XML") || mediaType.Contains("JSON") || mediaType.Contains("HTML"))
+                {
+                    if (content != null)
```
Comment (FermJacob) [inline Framework/BaseWebServiceTest/EventFiringWebServiceDriver.cs:276]: Add else for content being null to log?

**Review by FermJacob — CHANGES_REQUESTED** [review]

File: Framework/BaseWebServiceTest/EventFiringWebServiceDriver.cs (hunk 3 above)
Comment (TroyWalshProf) [reply Framework/BaseWebServiceTest/EventFiringWebServiceDriver.cs:282]: Added else

**Review by FermJacob — APPROVED** [review]

**Pull request merged**